docs/tutorials/ai/durable-ai-agent/index.md

```diff
@@ -0,0 +1,6078 @@
+)
+
+
+def generate_genai_prompt(
+    agent_goal: AgentGoal,
+    conversation_history: ConversationHistory,
+    raw_json: Optional[ToolData] = None,
+) -> str:
+    """
+    Generates a concise prompt for producing or validating JSON instructions
+    with the provided tools and conversation history.
+    """
+
+    # Prepare template variables
+    template_vars = {
+        "agent_goal": agent_goal,
+        "conversation_history_json": json.dumps(conversation_history, indent=2),
+        "toolchain_complete_guidance": TOOLCHAIN_COMPLETE_GUIDANCE_PROMPT,
+        "raw_json": raw_json,
+        "raw_json_str": (
+            json.dumps(raw_json, indent=2) if raw_json is not None else None
+        ),
+    }
+
+    return GENAI_PROMPT.render(**template_vars)
+
+
+def generate_tool_completion_prompt(current_tool: str, dynamic_result: dict) -> str:
+    """
+    Generates a prompt for handling tool completion and determining next steps.
+
+    Args:
+        current_tool: The name of the tool that just completed
+        dynamic_result: The result data from the tool execution
+
+    Returns:
+        str: A formatted prompt string for the agent to process the tool completion
+    """
+    return TOOL_COMPLETION_PROMPT.format(
+        current_tool=current_tool, dynamic_result=dynamic_result
+    )
+
+
+def generate_missing_args_prompt(
+    current_tool: str, tool_data: dict, missing_args: list[str]
+) -> str:
+    """
+    Generates a prompt for handling missing arguments for a tool.
+
+    Args:
+        current_tool: The name of the tool that needs arguments
+        tool_data: The current tool data containing the response
+        missing_args: List of argument names that are missing
+
+    Returns:
+        str: A formatted prompt string for requesting missing arguments
+    """
+    return MISSING_ARGS_PROMPT.format(
+        response=tool_data.get("response"),
+        current_tool=current_tool,
+        missing_args=missing_args,
+    )
+```
+</details>
+
+<details>
+
+<summary>
+Before moving on to the next section, verify your files and directory structure is correct.
+</summary>
+
+```
+temporal-ai-agent/
+├── .env.example
+├── .gitignore
+├── .python-version
+├── README.md
+├── pyproject.toml
+├── uv.lock
+├── activities/
+|   ├── __init__.py
+|   └── activities.py
+├── models/
+│   ├── __init__.py
+│   ├── core.py
+│   └── requests.py
+├── prompts/
+│   ├── __init__.py
+│   ├── agent_prompt_generators.py
+│   └── prompts.py
+├── scripts/
+│   ├── create_invoice_test.py
+│   ├── find_events_test.py
+│   └── search_flights_test.py
+└── tools/
+    ├── __init__.py
+    ├── create_invoice.py
+    ├── find_events.py
+    ├── goal_registry.py
+    ├── search_flights.py
+    ├── tool_registry.py
+    └── data/
+        └── find_events_data.json
+```
+</details>
+
+Now that you have the prompt rendering submodule implemented, you can implement the main agent Workflow.
+
+## Building the agent Workflow
```

Comment:
General comment -- this tutorial is *really long* and repetitive in places and this is the first thing approaching a real breakpoint, but it's not a great breakpoint because there's no end-to-end output that can be produced up to now. Can we do anything about that? I think we have to get this off of a single page somehow, I know our best examples of that (eg the background check in Go example) haven't been very successful at it either, but it's hard to imagine teaching this end-to-end right now, let alone someone dutifully following the whole thing on a single page.